### PR TITLE
(#1000) Created Timeout class

### PR DESCRIPTION
**Pull request by iakunin**
This is for #1000 

* Created class `Timeout`
* Created tests on `Timeout` class
* Removed puzzle

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1267](https://github.com/yegor256/cactoos/issues/1267) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1267](https://github.com/yegor256/cactoos/issues/1267) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), here is [why](https://www.0crat.com/footprint/C63314D6Z/54f77668-f548-46fb-a126-41e30e63000e); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by iakunin** [comment]
@fabriciofx could you have a look to this PR, please?

**Comment by paulodamaso** [comment]
@0crat assign me

File: src/main/java/org/cactoos/experimental/Timeout.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import org.cactoos.iterable.IterableEnvelope;
+import org.cactoos.iterable.IterableOf;
+
+/**
+ * Allows to execute the tasks concurrently within given timeout.
+ *
+ * @param <T> The type of task result item.
+ * @since 1.0.0
+ */
+public final class Timeout<T> extends IterableEnvelope<T> {
```
Comment (paulodamaso) [inline src/main/java/org/cactoos/experimental/Timeout.java]: @iakunin I think that `Timeout` does not represents correctly what this class is, maybe we should call it `Timed` or something like that, WDYT?

File: src/test/java/org/cactoos/experimental/TimeoutTest.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import org.junit.Test;
+import org.llorllale.cactoos.matchers.Assertion;
+import org.llorllale.cactoos.matchers.HasValues;
+import org.llorllale.cactoos.matchers.Throws;
+
+/**
+ * Test case for {@link Timeout}.
+ *
+ * @since 1.0.0
+ * @checkstyle MagicNumberCheck (500 lines)
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/experimental/TimeoutTest.java]: @iakunin We ca get rid of these check supressions using constants, please do it.

File: src/test/java/org/cactoos/experimental/TimeoutTest.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public final class TimeoutTest {
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/experimental/TimeoutTest.java]: @iakunin Also, put the first letter in the assertion strings in uppercase

File: src/test/java/org/cactoos/experimental/TimeoutTest.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+    /**
+     * Repeat the test several times.
+     * @param test The test to execute.
+     */
+    private void repeat(final org.cactoos.Proc<?> test) {
+        new UncheckedFunc<>(
+            new Repeated<>(
+                arg -> {
+                    test.exec(null);
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/experimental/TimeoutTest.java]: @iakunin What's that `null`  references?

**Comment by 0crat** [comment]
> @0crat assign me ([here](https://github.com/yegor256/cactoos/issues/1267#issuecomment-579021977))

@paulodamaso This pull request [#1267](https://github.com/yegor256/cactoos/issues/1267) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/C63314D6Z/fe79d52a-d49c-4433-9957-598f14e47eba); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by 0crat** [comment]
> @0crat assign me ([here](https://github.com/yegor256/cactoos/issues/1267#issuecomment-579021977))

@paulodamaso There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 11@542cb6aa2895, thread: PQ-C63314D6Z
com.zerocracy.claims.ClaimIn[214] java.lang.IllegalArgumentException: Parameter "age" not found in "Order was canceled" among: [comment, job, issue, repo, cause, _parent_sig, flow, login]
```

[0.51.0](https://github.com/zerocracy/farm): CID: [9bd2bc04-7a81-4755-872f-0a86e341c0b4](https://www.0crat.com/footprint/C63314D6Z/9bd2bc04-7a81-4755-872f-0a86e341c0b4), Type: "Order was canceled"

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

File: src/main/java/org/cactoos/experimental/Timeout.java (hunk 1 above)
Comment (iakunin) [reply src/main/java/org/cactoos/experimental/Timeout.java]: @paulodamaso you're right: I'll rename it to `Timed`

File: src/test/java/org/cactoos/experimental/TimeoutTest.java (hunk 2 above)
Comment (iakunin) [reply src/test/java/org/cactoos/experimental/TimeoutTest.java]: @paulodamaso sure. Done.

File: src/test/java/org/cactoos/experimental/TimeoutTest.java (hunk 3 above)
Comment (iakunin) [reply src/test/java/org/cactoos/experimental/TimeoutTest.java]: @iakunin sure. Done.

File: src/test/java/org/cactoos/experimental/TimeoutTest.java (hunk 4 above)
Comment (iakunin) [reply src/test/java/org/cactoos/experimental/TimeoutTest.java]: @paulodamaso got rid of this `null`

**Comment by iakunin** [comment]
@paulodamaso could you have a look one more time, please?

File: src/test/java/org/cactoos/experimental/TimedTest.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+import org.llorllale.cactoos.matchers.HasValues;
+import org.llorllale.cactoos.matchers.Throws;
+
+/**
+ * Test case for {@link Timed}.
+ *
+ * @since 1.0.0
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
```
Comment (paulodamaso) [inline src/test/java/org/cactoos/experimental/TimedTest.java]: @iakunin Seems that this supression is unnecessary now

Comment (iakunin) [reply src/test/java/org/cactoos/experimental/TimedTest.java]: @paulodamaso you're right. Removed, thanks.

File: src/test/java/org/cactoos/experimental/TimedTest.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            ).affirm()
+        );
+    }
+
+    /**
+     * Repeat the test several times.
+     * @param test The test to execute.
+     */
+    private void repeat(final Proc<Object> test) {
+        final Object dummy = new Object();
```
Comment (iakunin) [inline src/test/java/org/cactoos/experimental/TimedTest.java]: @paulodamaso also I've got an idea to create classes `RepeatedProc` and `RepeatedCallable` in order to get rid of this `dummy` objects. Also `Repeated` could be renamed to `RepeatedFunc`, but it'll break backward compatibility. 

I can create new issue or I can leave a puzzle in the scope of this issue. What do you think about it?

Comment (paulodamaso) [reply src/test/java/org/cactoos/experimental/TimedTest.java]: @iakunin Yes, I think that we should have `Repeated` decorator for use with `Proc` and `Callable`. Leave a puzzle for this, please.

Comment (iakunin) [reply src/test/java/org/cactoos/experimental/TimedTest.java]: @paulodamaso great! I've created a puzzle for this

**Comment by iakunin** [comment]
@paulodamaso see the puzzle, please, and get this PR finally be merged =)

**Comment by paulodamaso** [comment]
@iakunin Great, thanks!

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by 0crat** [comment]
Order was finished: +15 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Comment by 0crat** [comment]
Code review was too long (8 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso)

**Pull request merged**